kolibri/content/utils/channel_import.py

```diff
@@ -222,20 +239,72 @@ def table_import(self, model, row_mapper, table_mapper, unflushed_rows):
             self.destination.session.bulk_insert_mappings(DestinationRecord, data_to_insert)
         return unflushed_rows
 
+    def merge_record(self, data, model, DestinationRecord):
+        # Models that should be merged (see list above) need to be individually merged into the session
+        # as SQL Alchemy ORM does not support INSERT ... ON DUPLICATE KEY UPDATE style queries,
+        # as not available in SQLite, only MySQL as far as I can tell:
+        # http://hackthology.com/how-to-compile-mysqls-on-duplicate-key-update-in-sql-alchemy.html
+        RowEntry = self.destination.session.query(DestinationRecord).get(data[model._meta.pk.name])
+        if RowEntry:
+            for key, value in data.items():
+                setattr(RowEntry, key, value)
+        else:
+            RowEntry = DestinationRecord(**data)
+        self.destination.session.merge(RowEntry)
```

Comment:
If we're running through each record, is this a potential source of slowdown?

Reply:
Yes, there might be potential for optimisation here, but the mapping code needs to operate on the records in the current implementation. Possible we could encode that in direct sql queries.